woocommerce/woocommerce-ios
# [ABTest] Reload all experiments irrespective of logged in/out context

Closes: #8782 

## Description

**Problem**

User who was originally assigned `treatment` for a logged out variation sees a different variation for the same experiment after logging out.

**Reason for the change in assignment**
1. User launches app ABTest variations for logged out context experiments are fetched from the server and saved. 
1. After login due to Explat being reset the locally saved experiments are deleted.
1. We load the experiments for logged in context.
1. Now only logged in experiments have a variation assigned.
1. User logs out from Settings screen.
   1. [Due to change in `state`](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/Yosemite/DefaultStoresManager.swift#L196), `AppCoordinator` navigates to the login screen [here](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/ViewRelated/AppCoordinator.swift#L79).
   1. This navigation to login screen happens even before [this line](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/ViewRelated/Dashboard/Settings/Settings/SettingsViewController.swift#L433) from Settings is executed.
   1. The navigation happens also before analytics (i.e. ABTest) is reloaded [here](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/Yosemite/DefaultStoresManager.swift#L199).
1. When the login screen is displayed there is no variations assigned for the logged out experiment and `control` variant is shown to the user. 


**My fix in this PR**

We were fetching the variations for only specifc contexts (logged in/out) when analytics is reset. I changed that behaviour and instead started fetching the variations for all available experiments.

- The idea is to load the variations for all experiments irrespective of whether the user is logged in or not.
- As anonymous ID is assigned to requests sent in both logged in/out state [here](https://github.com/Automattic/Automattic-Tracks-iOS/blob/trunk/Sources/Experiments/ExPlatService.swift#L47), we are receiving the correct variation for logged out experiments even if the user is logged in/
- I believe the backend is able to use the [provided `anonID`](https://github.com/Automattic/Automattic-Tracks-iOS/blob/trunk/Sources/Experiments/ExPlatService.swift#L47) and [auth token](https://github.com/Automattic/Automattic-Tracks-iOS/blob/trunk/Sources/Experiments/ExPlatService.swift#L63) values to identify the device/user and provide correct variation values.

To verify this change works as expected
- [ ] @selanthiraiyan to add logged in and logged out A/A tests to verify the behaviour and check for crossovers.


**Better fix (Not in this PR)**

Wait for the ABTest variations to reload and proceed.
- Make [WooAnalytics.refreshUserData()](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/Analytics/WooAnalytics.swift#L58) async and wait for it to finish from everywhere we call the method.
- `WooAnalytics.refreshUserData()` method is currently called from
  - Upon switching store in [`SwitchStoreUseCase.finalizeStoreSelection()`](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/Authentication/Epilogue/SwitchStoreUseCase.swift#L135) and upon switching store in [`DefaultStoresManager.removeDefaultStore()`](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/Yosemite/DefaultStoresManager.swift#L182). I hope one of these calls can be avoided.
  - Upon deauthentication from [`DefaultStoresManager.deauthenticate()`](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/Yosemite/DefaultStoresManager.swift#L199)
  - Upon sync account from [`DefaultStoresManager.synchronizeAccount()`](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/Yosemite/DefaultStoresManager.swift#L279)
- As this solution requires changes to several critical areas I opted for the other solution for now.

## Testing instructions
- Install and launch the app. Logout if needed.
- Launch and skip onboarding.
- If you don't see the "Continue With WordPress.com" button to login you have been assigned `treatment` variation for the `woocommerceios_login_rest_api_project_202301_v2` experiment and you can test this PR (If not you need to use a sandbox to for a variation. You can follow the instructions from [this PR](https://github.com/woocommerce/woocommerce-ios/pull/8744) to assign a varition to your device.)
- Login into the app using site creds.
- Go to settings and logout.
- You will be navigated to the login screen and ensure that you still don't see "Continue With WordPress.com" button. i.e. You are still assigned the `treatment` variation.

## Screenshots
NA

---
